### PR TITLE
Added connection timeout

### PR DESCRIPTION
**Pull request by zhenlineo**
Default to 5s
Abstracted ConnectionSettings class to include all related settings for a connection inside and pass through to SocketConnection

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionSettings.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        public Uri InitialServerUri { get; }
+        public IAuthToken AuthToken { get; }
+        public TimeSpan ConnectionTimeout { get; }
+        public string UserAgent { get; }
+        public EncryptionManager EncryptionManager;
+
+        public ConnectionSettings(Uri initialServerUri, IAuthToken authToken, EncryptionManager encryptionManager, TimeSpan connectionTimeout, string userAgent = null)
+        {
+            Throw.ArgumentNullException.IfNull(initialServerUri, nameof(initialServerUri));
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionSettings.cs:35]: Could `Contract.Requires` be used here? As described here https://msdn.microsoft.com/en-us/library/dd264808(v=vs.110).aspx

File: Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs [hunk 2]
```diff
@@ -58,7 +71,11 @@ public void Init()
         {
             try
             {
-                Task.Run(() => _client.Start()).Wait();
+                var connected = Task.Run(() => _client.Start()).Wait(_connectionTimeout);
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs:74]: Shouldn't we cancel the `_client.Start()` task if `!connected`?
Could it still remain running in the background?
I do not know for sure, reading these articles:
https://msdn.microsoft.com/en-us/library/mt674886.aspx
https://msdn.microsoft.com/en-us/library/system.threading.cancellationtokensource(v=vs.110).aspx
https://msdn.microsoft.com/en-us/library/dd235606(v=vs.110).aspx

Would be nice to have this tested as well.

**Review by lutovich — CHANGES_REQUESTED** [review]
@zhenlineo review round completed

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionSettings.cs (hunk 1 above)
Comment (zhenlineo) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionSettings.cs:35]: I did not find this lib in .net core libs

File: Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs (hunk 2 above)
Comment (zhenlineo) [reply Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs:74]: We will close the whole connection (where we call `_client.Dispose()`) in `ConnectionPool.Acquire` if failed to create a new connection.

**Review by lutovich — APPROVED** [review]

**Pull request merged**